### PR TITLE
Add new REST APIs

### PR DESCRIPTION
**Pull request by yorma**
These APIs are added: 
1. create cachegroup
2. create server
3. edit server
4. assign deliveryservice servers
5. snapshot CRConfig
6. create profile
7. copy profile
8. create division
9. create region
10. create physical location.

### CODE REVIEW

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst [hunk 1]
```diff
@@ -244,3 +246,89 @@ Cache Group
...
+  +=================================+==========+===============================================================================+
+  | ``name``                        | yes      | The name of the Cache Group entry                                             |
+  +---------------------------------+----------+-------------------------------------------------------------------------------+
+  | ``short_name``                  | yes      | Abbreviation of the Cache Group Name                                          |
+  +---------------------------------+----------+-------------------------------------------------------------------------------+
+  | ``latitude``                    | yes      | Latitude for the Cache Group                                                  |
+  +---------------------------------+----------+-------------------------------------------------------------------------------+
+  | ``longitude``                   | yes      | Longitude for the Cache Group                                                 |
+  +---------------------------------+----------+-------------------------------------------------------------------------------+
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: parent_cachegroup and secondary_parent_cachegroup are not required

File: docs/source/development/traffic_ops_api/v12/profile.rst [hunk 2]
```diff
@@ -85,3 +85,93 @@ Profiles
...
+	Authentication Required: Yes
+
+	Role(s) Required: admin or oper
+
+	**Request Properties**
+
+	+-----------------------+--------+----------------------------------------------------+
+	|    Parameter          |  Type  |                    Description                     |
+	+=======================+========+====================================================+
+	| ``name``              | string | The name of the new profile copy to                |
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/profile.rst]: "copy to"? i think this description is wrong. this is not a copy, right?

File: docs/source/development/traffic_ops_api/v12/profile.rst [hunk 3]
```diff
@@ -85,3 +85,93 @@ Profiles
...
+|
+
+**POST /api/1.2/profiles/name/:profile_name/copy/:profile_copy_from**
+    Copy profile to an exist profile. 
+
+	Authentication Required: Yes
+
+	Role(s) Required: admin or oper
+
+	No Request Properties
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/profile.rst]: Please document the Request Route Parameters (:profile_name and :profile_copy_from) like this:

**Request Route Parameters**

Name | Required | Description
profile_name | yes | description
profile_copy_from | yes | description

File: docs/source/development/traffic_ops_api/v12/deliveryservice.rst [hunk 4]
```diff
@@ -1058,3 +1058,54 @@ SSL Keys
...
+
+**POST /api/1.2/deliveryservices/xmlId/:xml_id/servers**
+
+  Assign caches to a delivery service.
+
+  Authentication Required: Yes
+
+  Role(s) Required:  admin or oper
+
+  :mxl_id is the xml_id of the deliveryservice.
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/deliveryservice.rst]: to be consistent with the rest of the documentation, Document route parameters like

Request Route Parameters

Name | Required | Description
xml_id | yes | description

File: docs/source/development/traffic_ops_api/v12/phys_location.rst [hunk 5]
```diff
@@ -125,3 +125,106 @@ Physical Location
...
 |
+
+**POST /api/1.2/regions/:region_name/phys_locations**
+  Create physical location.
+
+  Authentication Required: Yes
+
+  Role(s) Required: admin or oper
+
+  region_name: the name of the region to create physical location into.
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/phys_location.rst]: to be consistent with the rest of the documentation, Document route parameters like

Request Route Parameters

Name | Required | Description
region_name | yes | description

File: docs/source/development/traffic_ops_api/v12/region.rst [hunk 6]
```diff
@@ -56,3 +56,55 @@ Regions
...
 |
+
+**POST /api/1.2/divisions/:division_name/regions**
+  Create Region
+
+  Authentication Required: Yes
+
+  Role(s) Required: admin or oper
+
+  division_name - The name of division to create new region into.
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/region.rst]: to be consistent with the rest of the documentation, Document route parameters like

Request Route Parameters

Name | Required | Description
division_name | yes | description

File: docs/source/development/traffic_ops_api/v12/server.rst [hunk 7]
```diff
@@ -440,3 +440,335 @@ Server
...
+|
+
+**PUT /api/1.2/servers/{:id}**
+
+  Allow user to edit server through api.
+
+  Authentication Required: Yes
+
+  Role(s) Required: admin or oper
+
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/server.rst]: to be consistent with the rest of the documentation, Document route parameters like

Request Route Parameters

Name | Required | Description
id | yes | description

File: docs/source/development/traffic_ops_api/v12/topology.rst [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+++++++++++++++++
+
+**PUT /api/1.2/snapshot/{:cdn_name}**
+
+  Authentication Required: Yes
+
+  Role(s) Required: admin or oper
+
+
+  {:cdn_name} -- name of the cdn to snapshot configure
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/topology.rst]: to be consistent with the rest of the documentation, Document route parameters like

Request Route Parameters

Name | Required | Description
cdn_name | yes | description

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst (hunk 1 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: @mitchell852 , there is parent_cachegroup_id and secondary_parent_cachegroup_id in cachegroup table. They should be required.
image

File: docs/source/development/traffic_ops_api/v12/profile.rst (hunk 2 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/profile.rst]: yes, will correct it. thanks.

File: docs/source/development/traffic_ops_api/v12/profile.rst (hunk 3 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/profile.rst]: ok

File: docs/source/development/traffic_ops_api/v12/deliveryservice.rst (hunk 4 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/deliveryservice.rst]: ok

File: docs/source/development/traffic_ops_api/v12/phys_location.rst (hunk 5 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/phys_location.rst]: ok

File: docs/source/development/traffic_ops_api/v12/region.rst (hunk 6 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/region.rst]: ok

File: docs/source/development/traffic_ops_api/v12/server.rst (hunk 7 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/server.rst]: ok

File: docs/source/development/traffic_ops_api/v12/topology.rst (hunk 8 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/topology.rst]: ok

**Comment by yorma** [comment]
@mitchell852 I make changes as your comments, except parent_cachegroup and secondary_parent_cachegroup, since I think parent_cachegroup and secondary_parent_cachegroup  are requried in the API. Thanks,

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst (hunk 1 above)
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: @dangogh ? thoughts?

Comment (dangogh) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: Nor required.  *Some* cachegroup needs to have no parent or there would be cycles.

File: traffic_ops/app/lib/API/Server.pm [hunk 9]
```diff
@@ -364,4 +364,405 @@ sub details {
...
+
+    if ( defined( $json->{'tcp_port'} ) ) {
+        $params{'tcp_port'} = int( $json->{'tcp_port'} );
+    }
+    elsif ($flag_create) {
+        $params{'tcp_port'} = 80;
+    }
+
+    if ( defined( $json->{'cachegroup'} ) ) {
+        eval {
```
Comment (dangogh) [inline traffic_ops/app/lib/API/Server.pm]: is eval really necessary here?  It incurs extra overhead,  and checking that the return value is defined should be sufficient.   Same goes for all the eval's I see here..

File: traffic_ops/app/lib/API/Cachegroup.pm [hunk 10]
```diff
@@ -144,4 +147,78 @@ sub available_for_parameter {
...
+    my $params = $self->req->json;
+    if (!defined($params)) {
+        return $self->alert("parameters must Json format,  please check!"); 
+    }
+
+    if ( !&is_oper($self) ) {
+        return $self->alert("You must be an ADMIN or OPER to perform this operation!");
+    }
+
+    my $cachegroups = &UI::Cachegroup::get_cachegroups($self);
```
Comment (dangogh) [inline traffic_ops/app/lib/API/Cachegroup.pm]: get_cachegroups is a small and simple method,  and we'd rather not introduce dependencies on the UI modules for something like that.   I'd rather see that method copied here..

File: traffic_ops/app/lib/API/DeliveryService.pm [hunk 11]
```diff
@@ -27,6 +27,10 @@ use Email::Valid;
 use Validate::Tiny ':all';
 use Data::Dumper;
 use Common::ReturnCodes qw(SUCCESS ERROR);
+use JSON;
+use MojoPlugins::Response;
+use UI::DeliveryService;
+use UI::Topology;
```
Comment (dangogh) [inline traffic_ops/app/lib/API/DeliveryService.pm]: another instance of introducing dependency on UI,  but header_rewrite code is more complex than previous example... not sure what to do with that...

Is UI::Topology required for UI::DeliveryService?  If not, can it be removed?

File: traffic_ops/app/lib/API/DeliveryService.pm [hunk 12]
```diff
@@ -392,7 +396,62 @@ sub is_deliveryservice_request_valid {
 	else {
 		return ( 0, $result->{error} );
 	}
+}
+
+sub assign_servers {
+	my $self   = shift;
+    my $ds_xml_Id = $self->param('xml_id');
```
Comment (dangogh) [inline traffic_ops/app/lib/API/DeliveryService.pm]: minor indentation issue..   perltidy?

File: traffic_ops/app/lib/API/PhysLocation.pm [hunk 13]
```diff
@@ -71,4 +73,79 @@ sub index_trimmed {
...
+
+    if ( !defined($default) ) {
+        return $v;
+    }
+    if ( !defined($v) ) {
+        return $default;
+    }
+    return $v;
+}
+
```
Comment (dangogh) [inline traffic_ops/app/lib/API/PhysLocation.pm]: ```return $v // $default;``` should have the same effect..

File: traffic_ops/app/lib/API/Profile.pm [hunk 14]
```diff
@@ -59,6 +59,93 @@ sub index_trimmed {
...
+	my $self = shift;
+	my $params = $self->req->json;
+	if ( !defined($params) ) {
+		return $self->alert("parameters must Json format,  please check!");
+	}
+
+    if ( !&is_oper($self) ) {
+        return $self->alert( { Error => " - You must be an admin or oper to perform this operation!" } );
+    }
+
```
Comment (dangogh) [inline traffic_ops/app/lib/API/Profile.pm]: indentation -- perltidy

File: traffic_ops/app/t/api/1.2/cachegroup.t [hunk 15]
```diff
@@ -0,0 +1,93 @@
+package main;
```
Comment (dangogh) [inline traffic_ops/app/t/api/1.2/cachegroup.t]: *very* helpful that tests are included!

File: traffic_ops/app/lib/API/Server.pm (hunk 9 above)
Comment (yorma) [inline traffic_ops/app/lib/API/Server.pm]: @dangogh, It is to check if json parameters correct or not. If not pass check will return corresponding error message to caller. For some parameters if caller omits, will give default value. These kind of check can give caller clear messages if the input parameter is wrong. I prefer to keep current checking for parameter value.

File: traffic_ops/app/lib/API/Cachegroup.pm (hunk 10 above)
Comment (yorma) [inline traffic_ops/app/lib/API/Cachegroup.pm]: ok, will make a copy of get_cachegroups().

File: traffic_ops/app/lib/API/DeliveryService.pm (hunk 11 above)
Comment (yorma) [inline traffic_ops/app/lib/API/DeliveryService.pm]: will remove UI::Topology, thanks.

File: traffic_ops/app/lib/API/DeliveryService.pm (hunk 12 above)
Comment (yorma) [inline traffic_ops/app/lib/API/DeliveryService.pm]: ok

File: traffic_ops/app/lib/API/PhysLocation.pm (hunk 13 above)
Comment (yorma) [inline traffic_ops/app/lib/API/PhysLocation.pm]: got it, thanks.

File: traffic_ops/app/lib/API/Profile.pm (hunk 14 above)
Comment (yorma) [inline traffic_ops/app/lib/API/Profile.pm]: ok

**Comment by yorma** [comment]
@dangogh @mitchell852 , I modified with your comments, please check. Thanks.

**Comment by yorma** [comment]
@dangogh @mitchell852 ,  I resolved conflicts because there are other pull requests merged to master. How the status of this pull request? thanks

**Comment by dangogh** [comment]
unfortunately,   it looks like your merge didn’t go so well.   Your PR now appears to include ​*all*​ the changes that have been submitted,  and so it’s hard to tell what changes you actually made and whether the changes others made are included in the  result.  That makes it really difficult to evaluate this change.     I suggest you go back to the version before this merge (git reset --soft <commit number>) and use "git rebase -i upstream master" to make sure only your changes are included in the PR.   I'm happy to try to help walk you through that if you have any trouble..   -Dan

**Comment by yorma** [comment]
@dangogh @mitchell852 , my fault in git rebase, thanks for pointing out. I cherry-pick to another branch and create a new PR https://github.com/Comcast/traffic_control/pull/1142 . Let's use new PR, thanks.

**Pull request closed without merging**